content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+
+```
+(gdb) gcore
+Saved corefile core.1628
+(gdb) quit
+
+= du -h core.1628
+1.2G    core.1628
+```
+
+With coredump saved, we're no longer required to keep the process running. Feel free to “kill -9” it.
+
+*For more info on the topic of debugging with GDB, check out Go’s own "[Debugging Go Code with GDB][2]".*
+
+## Enter Delve, Debugger for Go
+
+[Delve][] is the debugger for Go programs. It is similar to GDB but is aware of Go's runtime, data structures and other internals of the language.
+
+I highly recommend a talk by Alessandro Arzilli “[Internal Architecture of Delve, a Debugger For Go](https://www.youtube.com/watch?v=IKnTr7Zms1k)” from GopherCon EU 2018 if you're interested to know about Delve's internals.
+
+Delve is written in Go so installing it as simple as running:
+
+```
+= go get -u github.com/derekparker/delve/cmd/dlv
+```
+
+With Delve installed we can start analysing the core file by running `dlv core <path to service binary> <core file>`. We start by listing all goroutines that were running when the coredump was taken. Delve's `goroutines` command does exactly this:
+
+```
+= dlv core example/server/server core.1628
+
+(dlv) goroutines
+  ···
+  Goroutine 4611 - User: /vagrant/example/server/metrics.go:113 main.(*Metrics).CountS (0x703948)
+  Goroutine 4612 - User: /vagrant/example/server/metrics.go:113 main.(*Metrics).CountS (0x703948)
+  Goroutine 4613 - User: /vagrant/example/server/metrics.go:113 main.(*Metrics).CountS (0x703948)
+
+```
+
+*Unfortunately, in a real case scenario, the list can be so big, it doesn't even fit into terminal's scroll buffer. Remember that the server spawns a goroutine for each incoming request, so “goroutines” command has shown us a list of almost a million items. Let's pretend that we faced exactly this, and think of a way to work through this situation.*
+
+We can run Delve in the "headless" mode to interact with the debugger via it's [JSON-RPC API](https://github.com/derekparker/delve/tree/master/Documentation/api).
+
+Run the same `dlv core` command as we previously did, but this time specify that we need to start Delve’s API server:
+
+```
+= dlv core example/server/server core.1628 --listen :44441 --headless --log
+API server listening at: [::]:44441
+INFO[0000] opening core file core.1628 (executable example/server/server)  layer=debugger
+```
+
+After debug server is running, we can send commands to it’s TCP port and store the output as raw JSON. Let's get the list of running goroutines once again, but this time saving the results to a file:
+
+```
+= echo -n '{"method":"RPCServer.ListGoroutines","params":[],"id":2}' | nc localhost 44441 > server-test-1_dlv-rpc-list_goroutines.json
```

Comment:
Add `-w 1` to the netcat command so that it will terminate one second after EOF from stdin, otherwise the command will not terminate at all
```suggestion
= echo -n '{"method":"RPCServer.ListGoroutines","params":[],"id":2}' | nc -w 1 localhost 44441 > server-test-1_dlv-rpc-list_goroutines.json
```